libraries/AP_AngleSensor/AP_AngleSensor.cpp

```diff
@@ -0,0 +1,177 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AP_AngleSensor.h"
+
+#if AP_ANGLESENSOR_ENABLED
+
+#include "AP_AngleSensor_AS5048B.h"
+#include <AP_Logger/AP_Logger.h>
+#include <AP_Math/AP_Math.h>
+#include <GCS_MAVLink/GCS.h>
+
+extern const AP_HAL::HAL& hal;
+
+const AP_Param::GroupInfo AP_AngleSensor::var_info[] = {
+    // @Group: _
+    // @Path: AP_AngleSensor_Params.cpp
+    AP_SUBGROUPINFO(_params[0], "_", 0, AP_AngleSensor, AP_AngleSensor_Params),
+
+    #if ANGLE_SENSOR_MAX_INSTANCES > 1
+    // @Group: 2_
+    // @Path: AP_AngleSensor_Params.cpp
+    AP_SUBGROUPINFO(_params[1], "2_", 1, AP_AngleSensor, AP_AngleSensor_Params),
+    #endif
+    AP_GROUPEND
+};
+
+AP_AngleSensor::AP_AngleSensor(void)
+{
+    _singleton = this;
+
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+// initialise the AP_AngleSensor class.
+void AP_AngleSensor::init(void)
+{
+    if (_num_instances != 0) {
+        // init called a 2nd time?
+        return;
+    }
+    for (uint8_t i=0; i<ANGLE_SENSOR_MAX_INSTANCES; i++) {
+        switch ((AngleSensor_Type)_params[i]._type.get()) {
+
+        case ANGLESENSOR_TYPE_AS5048B:
+            drivers[i] = new AP_AngleSensor_AS5048B(*this, i, state[i]);
+            break;
+            
+        case ANGLESENSOR_TYPE_NONE:
+            break;
+        }
+
+        if (drivers[i] != nullptr) {
+            // we loaded a driver for this instance, so it must be
+            // present (although it may not be healthy)
+            _num_instances = i+1;  // num_instances is a high-water-mark
+        }
+    }
+}
+
+// update Angle Sensor state for all instances. This should be called by main loop
+void AP_AngleSensor::update(void)
+{
+    for (uint8_t i=0; i<_num_instances; i++) {
+        if (drivers[i] != nullptr && _params[i]._type != ANGLESENSOR_TYPE_NONE) {
```

Comment:
I presume the type check is so you can runtime disable? But that would just leave the state unchanged which could still be healthy?

Better to either not allow runtime disable, or if disabled then overwrite the state.